Add ORC writer flush policy

sdruzkin opened the pull request:
Extract current flush policy into a separate object behind an interface to be able to customize flush policy.

Flush policy should be implemented as a new OrcWriterFlushPolicy interface. Existing logic is extracted into DefaultOrcWriterFlushPolicy. The default implementation is exactly the same as the replaced logic. OrcWriterFlushPolicy passed to the OrcWriter as a part of OrcWriterOptions. 

Unfortunately due to tight coupling with the Dictionary Optimizer it's almost impossible to neatly encapsulate flush decision without exposing stripe min/max size and max stripe row count, so I let it be in the flush policy interface.

Test plan:
* added new unit tests
* ran existing unit tests
* completed verifier run with id 75500
   * `Progress: 9527 succeeded, 451 skipped, 0 resolved, 0 failed, 100.00% done` 

```
== NO RELEASE NOTE ==
```

sdruzkin commented on presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java:
```diff
@@ -46,11 +47,15 @@
-        // Give separate copy to callers for isolation.
-        return OrcWriterOptions.builder()
+        DefaultOrcWriterFlushPolicy flushPolicy = DefaultOrcWriterFlushPolicy.builder()
                 .withStripeMinSize(stripeMinSize)
                 .withStripeMaxSize(stripeMaxSize)
                 .withStripeMaxRowCount(stripeMaxRowCount)
```
These settings moved to DefaultOrcWriterFlushPolicy

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/DefaultOrcWriterFlushPolicy.java:
```diff
@@ -0,0 +1,138 @@
+public class DefaultOrcWriterFlushPolicy
+        implements OrcWriterFlushPolicy
+{
+    public static final DataSize DEFAULT_STRIPE_MIN_SIZE = new DataSize(32, MEGABYTE);
```
These constants have been moved from OrcWriterOptions

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/DefaultOrcWriterFlushPolicy.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public Optional<FlushReason> shouldFlushStripe(int stripeRowCount, int bufferedBytes, boolean dictionaryIsFull)
+    {
+        if (stripeRowCount >= stripeMaxRowCount) {
```
This is the existing behavior except strict `>` were replaced with `>=`

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/DefaultOrcWriterFlushPolicy.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public int getStripeMinBytes()
```
These getters are used by the dictionary optimizer. I was not able to nicely encapsulate them.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/FlushReason.java:
```diff
@@ -0,0 +1,19 @@
+ */
+package com.facebook.presto.orc;
+
+public enum FlushReason
```
This class was extracted from WriterStats without any changes.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java:
```diff
@@ -637,7 +620,7 @@ private void flushColumnWriters(FlushReason flushReason)
         closedStripesRetainedBytes += closedStripe.getRetainedSizeInBytes();
 
         recordValidation(validation -> validation.addStripe(stripeInformation.getNumberOfRows()));
-        stats.recordStripeWritten(stripeMinBytes, stripeMaxBytes, dictionaryMaxMemoryBytes, flushReason, dictionaryCompressionOptimizer.getDictionaryMemoryBytes(), stripeInformation);
```
stripeMinBytes, stripeMaxBytes, dictionaryMaxMemoryBytes values were not used by stats, so I removed them.

arunthirupathi replied:
Let us keep passing this value. We need to record this for Jolly when our stats implementation is ready.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:
```diff
@@ -45,9 +42,7 @@
     public static final DwrfStripeCacheMode DEFAULT_DWRF_STRIPE_CACHE_MODE = INDEX_AND_FOOTER;
     public static final int DEFAULT_PRESERVE_DIRECT_ENCODING_STRIPE_COUNT = 0;
 
-    private final DataSize stripeMinSize;
```
These fields moved to OrcWriterFlushPolicy. A new flushPolicy field was added.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterStats.java:
```diff
@@ -37,9 +37,6 @@
 
     @Override
     public void recordStripeWritten(
-            int stripeMinBytes,
```
These values were not used by stats, so I removed them.

sdruzkin replied:
Reverted changes to OrcWriterStats

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/DefaultOrcWriterFlushPolicy.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public Optional<FlushReason> shouldFlushStripe(int stripeRowCount, int bufferedBytes, boolean dictionaryIsFull)
```
When a custom flush policy wants to flush, what FlushReason can it provide ? I recommend adding CUSTOM or changing this to some unstructured type like String.

sdruzkin replied:
I'll add OTHER. 

As far as I can see it's only used in stats, and when the flush reason is CLOSED - which should probably be modeled in a different way. So making FlushReason object a bit more flexible doesn't make much sense.

arunthirupathi reviewed (commented):
Did you check the file size difference for the Validation run with this change ? I will ping you offline with a dashboard on how to perform the file size difference analysis.

sdruzkin commented:
> Did you check the file size difference for the Validation run with this change ? I will ping you offline with a dashboard on how to perform the file size difference analysis.

There should be zero to none difference because the flush border (getChunkMaxLogicalBytes) remained unchanged. You know what, there is no reason to change default flush conditions - I'll change the flush condition back to original `strict greater than`.

arunthirupathi commented:
> There should be zero to none difference because the flush border (getChunkMaxLogicalBytes) remained unchanged. You know what, there is no reason to change default flush conditions - I'll change the flush condition back to original `strict greater than`.

I believe the change is fine, verifying the file sizes is a sanity check to ensure that, we are not regressing anything unintentionally. (Say a bad default). I don't see any issues at all in the code though.

sdruzkin commented:
Btw, @arunthirupathi  what do you think about moving this piece from writer into the flush policy?
```
        // avoid chunk with huge logical size
        double averageLogicalSizePerRow = (double) page.getApproximateLogicalSizeInBytes() / page.getPositionCount();
        int maxChunkRowCount = max(1, (int) (flushPolicy.getChunkMaxLogicalBytes() / max(1, averageLogicalSizePerRow)));
```
will turn into
```
 int maxChunkRowCount = flushPolicy.getChunkRowCount(page);
```

This will make flush check more flexible by introducing an option to create write chunks either by row count or by chunk logical size (current behavior).

sdruzkin commented:
Rev 2: 
* reverted changes to WriterStats.recordStripeWritten
* made flush logic exactly as it was originally
* added CUSTOM flush reason
* made FlushPolicy return the number of rows between flush checks (getMaxChunkRowCount)

arunthirupathi commented:
> Btw, @arunthirupathi what do you think about moving this piece from writer into the flush policy?
> 
> ```
>         // avoid chunk with huge logical size
>         double averageLogicalSizePerRow = (double) page.getApproximateLogicalSizeInBytes() / page.getPositionCount();
>         int maxChunkRowCount = max(1, (int) (flushPolicy.getChunkMaxLogicalBytes() / max(1, averageLogicalSizePerRow)));
> ```
> 
> will turn into
> 
> ```
>  int maxChunkRowCount = flushPolicy.getChunkRowCount(page);
> ```
> 
> This will make flush check more flexible by introducing an option to create write chunks either by row count or by chunk logical size (current behavior).

I believe this logic should not be moved, this will let flush policy control the row count.

sdruzkin commented:
Rev 3: forgot to adjust the new unit test to the reverting of the logic to the original behavior

sdruzkin commented:
I see some random failures in presto-tests. Re-triggering tests.

sdruzkin commented:
All checks passed.
